R/enumerate_all_paths.R

```diff
@@ -0,0 +1,160 @@
+enumerate_all_paths <- function(dm, start) {
+  all_fks <-
+    dm %>%
+    dm_get_all_fks() %>%
+    rename(child_cols = child_fk_cols, parent_cols = parent_key_cols)
+
+  graph_df_ud <-
+    all_fks %>%
+    rename(
+      child_table = parent_table,
+      child_cols = parent_cols,
+      parent_table = child_table,
+      parent_cols = child_cols,
+    ) %>%
+    bind_rows(all_fks, .)
+
+  helper_env <- new_environment()
+  assign("tbl_node", list(), envir = helper_env)
+  assign("all_paths", tibble(
+    child_table = character(),
+    child_cols = new_keys(),
+    parent_table = character(),
+    parent_cols = new_keys(),
+    new_child_table = character(),
+    new_parent_table = character()
+  ), envir = helper_env)
+  enumerate_all_paths_impl(start, graph_df_ud = graph_df_ud, helper_env = helper_env)
+  all_paths <- get("all_paths", envir = helper_env)
+  # need to take into account FKs from unconnected components (graph representation)
+  fks_from_unconnected <- anti_join(
+    all_fks,
+    all_paths,
+    by = c("child_table", "parent_table")
+  ) %>%
+    mutate(new_child_table = child_table, new_parent_table = parent_table)
+  all_paths %>%
+    rename_unique() %>%
+    bind_rows(fks_from_unconnected) %>%
+    split_to_list()
+}
+
+enumerate_all_paths_impl <- function(node,
+                                     node_key_cols = character(),
+                                     former_node = character(),
+                                     former_key_cols = character(),
+                                     path = character(),
+                                     graph_df_ud,
+                                     helper_env) {
+  if (length(path) > 0) {
+    # increase tbl_node[[node]] by 1, return this index in a suffix
+    usage_idx <- inc_tbl_node(node, helper_env)
+    add_path_to_all_paths(
+      graph_df_ud,
+      node,
+      node_key_cols,
+      former_node,
+      former_key_cols,
+      new_former_node = names(path)[[length(path)]],
+      usage_idx,
+      helper_env
+    )
+  } else {
+    usage_idx <- ""
+  }
+
+  path <- c(path, set_names(node, paste0(node, usage_idx)))
+  out <-
+    graph_df_ud %>%
+    filter(child_table == !!node) %>%
+    filter(!(parent_table %in% path)) %>%
+    rename(node = parent_table, node_key_cols = parent_cols, former_node = child_table, former_key_cols = child_cols) %>%
+    select(-on_delete)
+
+  if (nrow(out) == 0) {
+    return()
+  }
+
+  pwalk(out, enumerate_all_paths_impl, path, graph_df_ud, helper_env)
+}
+
+rename_unique <- function(all_paths) {
+  all_names <- bind_rows(
+    select(all_paths, table = child_table, new_table = new_child_table),
+    select(all_paths, table = parent_table, new_table = new_parent_table)
+  ) %>%
+    distinct() %>%
+    arrange(table, new_table) %>%
+    group_by(table) %>%
+    count()
```

Comment:
This is shorter and will also result in an ungrouped data frame:

```suggestion
    count(table)
```